### ISSUE TITLE
the inputs of the "refinement model" are different between train.py and test.py

### ISSUE THREAD

**Issue by vince2003** [VinAIResearch/MagNet#8]
Hello @hmchuong,
Thank you for your source code. I read your code and recognize that the inputs of the "refinement model" between train.py and test.py are different. To more detail, the 2 inputs of the "refinement model" in train.py (crop_preds, fine_pred) are derived from the backbone model while the 2 inputs of the "refinement model" in test.py (scale_early_preds, coarse_preds) are derived from the backbone model and the refinement model of the previous stage, respectively. 

Can you explain why the inputs of the "refinement model" are different between train.py and test.py? Thank you very much!

# In train.py
coarse_pred = model(coarse_image).softmax(1) # "model" is backbone model
fine_pred = model(fine_image).softmax(1)
crop_preds = roi_align(coarse_pred, coords, output_size=(opt.input_size[1], opt.input_size[0]))
logits = refinement_model(crop_preds, fine_pred) # --> "crop_preds" and "fine_pred" derived from the backbone model

#--------------------
# In test.py
scale_early_preds = get_batch_predictions(model, sub_batch_size, scale_image_patches.to(device))
coarse_preds = roi_align(final_output, [coords[selected_patch_ids]], output_size=(opt.input_size[1], opt.input_size[0])) # "final_output" derived from the refinement model --> "coarse_preds" derived from the refinement model
.
.
fine_pred = get_batch_predictions(
    refinement_models[min(len(refinement_models), idx) - 1],
    sub_batch_size,
    scale_early_preds, # --> "scale_early_preds" from backbone model
    coarse_preds, #---> "coarse_preds" from the refinement model of the previous stage
)
.
.
final_output = (
    final_output.reshape(1, opt.num_classes, scale[0] * scale[1])
    .scatter_(2, error_point_indices, fine_pred) # "fine_pred" derived from the refinement model
    .view(1, opt.num_classes, scale[1], scale[0])
)

**Comment by hmchuong** [comment]
Hi, as mentioned in the paper, to speed up the training process, the refinement module was trained to refine the whole output from the backbone network. In inference or testing, we will use the trained refinement module to correct progressively the about from preceding stages.

**Comment by vince2003** [comment]
Hi, Thank you for your reply. However, I don't understand why the refinement model is trained on the input source (the output of the backbone model) while it is tested on the different input sources (the output of the preceding stages of the refinement model). So, these inputs have different distributions. With regarding IOU performance, Can it work in this way? What is the purpose of training refinement model? 
Can you explain them to me? Thank you!

**Comment by hmchuong** [comment]
I think you misunderstood some points. The testing is still performed on the output of the backbone. After the refinement step, we only refine a part of the output of the model. The output is then used for the next stage.

**Comment by vince2003** [comment]
Thank you for your answer!
Of course, The testing is still performed on the output of the backbone. Sorry, my question makes you confuse. I will simplify my question as follows: 
## in train.py 
Output= refinement(A,B) 
## in testing.py 
Output= refinement(A,C) 

- A is the output of the backbone. 
- B is cropped from the output of the backbone. 
- C is the output of the refinement model of the previous stage.
So, B and C are totally different. Is my understanding true? Is it ok if B and C are different? 
Sorry for my limited understanding. 
Thank you!

**Comment by hmchuong** [comment]
C is not totally the output of the refinement model of the previous stage. The output from the backbone (O) in the first stage is kept through stages. At each stage, a small portion of the output O is replaced by the output of the refinement module (to correct errors).

**Comment by quangvinh5** [comment]
Thank you for your answer!

**Comment by vince2003** [comment]
Thank you for your answer!
